napari/layers/base/base.py

```diff
@@ -999,6 +1009,131 @@ def _data_to_world(self) -> Affine:
         """
         return self._transforms[1:3].simplified
 
+    def vector_world_to_data(self, vector) -> tuple:
+        """Convert a vector defining an orientation from world coordinates to data coordinates.
+        For example, this would be used to convert the view ray.
+
+        Parameters
+        ----------
+        vector : tuple, list, 1D array
+            A vector in world coordinates.
+
+        Returns
+        -------
+        tuple
+            Vector in data coordinates.
+        """
+        vector = np.asarray(vector)
+
+        # create transform for view direction (world -> layer data)
+        shear_matrix = expand_upper_triangular(self.shear)
+        rot_matrix = self.rotate
+        scale_matrix = np.diag(self.scale)
+        world_to_layer = np.linalg.inv(
+            shear_matrix @ rot_matrix
+        ).T @ np.linalg.inv(scale_matrix)
+
+        return tuple(vector @ world_to_layer)
+
+    @property
+    def _display_bounding_box(self):
+        """An axis aligned (self._ndisplay, 2) bounding box around the data"""
+        return self._extent_data[:, self._dims_displayed].T
+
+    def _data_face_intersection(
+        self,
+        vertices: np.ndarray,
+        click_pos_data: np.ndarray,
+        view_dir_data: np.ndarray,
+    ):
+
+        # project the vertices on to the view plane
+        vertices_plane = project_point_to_plane(
+            point=vertices,
+            plane_point=click_pos_data,
+            plane_normal=view_dir_data,
+        )
+
+        # rotate the plane to make the triangles 2D
+        rotation_matrix = rotation_matrix_from_vectors(
+            view_dir_data, [0, 0, 1]
+        )
+        rotated_vertices = vertices_plane @ rotation_matrix.T
+        vertices_2D = rotated_vertices[:, :2]
+        click_pos_2D = rotation_matrix.dot(click_pos_data)[:2]
+
+        triangle_vertices_2D = np.stack(
+            (vertices_2D[[0, 1, 2]], vertices_2D[[0, 2, 3]])
+        )
+        in_triangles = inside_triangles(triangle_vertices_2D - click_pos_2D)
+        if in_triangles.sum() > 0:
+            return True
+        else:
+            return False
+
+    def _cursor_ray(self, event):
```

Comment:
wow, super cool - i'll want to take a closer look at this to understand before merge, but one thought is if it might be possible to pull a little more logic out into utils functions so the function on the base layer is just a couple key calls to those utils (i know you've already done that quite a bit!)

Reply:
Yes, we definitely should refactor this a bit. A lot of the code here is actually pretty general geometry stuff that should be pulled out into the geometry utils (e.g., the `_data_face_intersection()` method you flagged above).

Reply:
I've refactored the interaction code quite a bit and I think it's a bit easier to read now.